python/dolfinx/nls/petsc.py

```diff
@@ -72,3 +98,418 @@ def setP(self, P: types.FunctionType, Pmat: PETSc.Mat):  # type: ignore
+    for F_sub, bc, x_sub in zip(F.getNestSubVecs(), bcs0, sub_vectors):
+        set_bc(F_sub, bc, x_sub, -1.0)
+
+    # Must assemble F here in the case of nest matrices
+    F.assemble()
+
+
+def J_nest(
+    u: list[dolfinx.fem.Function],
+    jacobian: list[list[dolfinx.fem.Form]],
+    preconditioner: typing.Optional[list[dolfinx.fem.Form]],
+    bcs: list[dolfinx.fem.DirichletBC],
+    snes: PETSc.SNES,  # type: ignore
+    x: PETSc.Vec,  # type: ignore
+    J: PETSc.Mat,  # type: ignore
+    P: PETSc.Mat,  # type: ignore
+):
+    """Assemble the Jacobian matrix.
+
+    Args:
+        u: List of functions tied to the solution vector within the residual and jacobian
+        jacobian: List of list of forms of the Jacobian
+        preconditioner: List of forms of the preconditioner
+        bcs: List of Dirichlet boundary conditions
+        snes: The solver instance
+        x: The vector containing the latest solution
+        J: Matrix to assemble the Jacobian into
+        P: Matrix to assemble the preconditioner into
+    """
+    copy_nest_vec_to_functions(u, x)
+    assert J.getType() == "nest" and P.getType() == "nest"
+    J.zeroEntries()
+    assemble_matrix_nest(J, jacobian, bcs=bcs, diagonal=1.0)  # type: ignore
+    J.assemble()
+
+    if preconditioner is not None:
+        P.zeroEntries()
+        assemble_matrix_nest(P, preconditioner, bcs=bcs, diagonal=1.0)  # type: ignore
+        P.assemble()
+
+
+def create_snes_solver(
+    F: typing.Union[dolfinx.fem.Form, ufl.form.Form],
+    u: dolfinx.fem.Function,
+    bcs: typing.Optional[list[dolfinx.fem.DirichletBC]] = None,
+    J: typing.Optional[typing.Union[dolfinx.fem.Form, ufl.form.Form]] = None,
+    P: typing.Optional[typing.Union[dolfinx.fem.Form, ufl.form.Form]] = None,
+    assembly_type: fem.AssemblyType = fem.AssemblyType.default,
+    form_compiler_options: typing.Optional[dict] = None,
+    jit_options: typing.Optional[dict] = None,
+) -> tuple[PETSc.SNES, PETSc.Vec]:  # type: ignore
+    """Create a PETSc SNES solver instance and a vector to store solutions in.
+
+    Solves problems of the form :math:`F_i(u, v) = 0, i=0,...N\\ \\forall v \\in V`
+    where :math:`u=(u_0,...,u_N), v=(v_0,...,v_N)` using PETSc SNES as the non-linear solver.
+
+    Note:
+        The user is responsible for the destruction of the returned objects.
+
+    Args:
+        F: List of PDE residuals :math:`F_i`.
+        u: List of unknowns of the block system.
+        bcs: List of Dirichlet boundary conditions.
+        J: Rectangular array of bi-linear forms representing the
+            Jacobian :math:`J_ij = dF_i/du_j`.
+        P: Rectangular array of bi-linear forms representing the preconditioner.
+        assembly_type: Determines what kind of matrix the problem is assembled into.
+            If `N=0`, i.e one is solving for a single form, the assembly type is `default`.
+            For blocked, the assembly type is `block` or `nest`.
```

Comment:
I wonder if we could just make default mean non-blocked or blocked, depending on if F is a Form or a list of Forms.

Reply:
You would still need an argument to determine nest or not, so Im not sure if it gives much.